### PR TITLE
Add flags for GNU builds to allow using GNU v10 and higher versions

### PR DESCRIPTION
**Pull request by ndkeen**
Add -fallow-argument-mismatch for all GNU builds (including gnugpu).

Only add -fallow-invalid-boz for 2 specific files for gnu/gnugpu builds.

This PR adds Depends.gnugpu.cmake which is similar to Depends.gnu.cmake

Fixes https://github.com/E3SM-Project/E3SM/issues/3954
And is needed first for https://github.com/E3SM-Project/E3SM/issues/4809 and https://github.com/E3SM-Project/E3SM/pull/4818

[BFB]

### CODE REVIEW

File: cime_config/machines/Depends.gnugpu.cmake [hunk 1]
```diff
@@ -0,0 +1,24 @@
+list(APPEND MPAS_ICE_SHORTWAVE
```
Comment (jayeshkrishna) [inline cime_config/machines/Depends.gnugpu.cmake:1]: Following the patterns of the other changes in this PR you might want to rename the list to DISABLE_TREE_PTA_LIST or something like that

Comment (jayeshkrishna) [reply cime_config/machines/Depends.gnugpu.cmake:1]: And you might want to amend the commit message to include this flag

Comment (ndkeen) [reply cime_config/machines/Depends.gnugpu.cmake:1]: The MPAS_ICE_SHORTWAVE name was already there. It's fine to rename it...
I don't know what you mean about amending commit message

Comment (jayeshkrishna) [reply cime_config/machines/Depends.gnugpu.cmake:1]: The commit message (or PR desc) does not mention the new compiler flag, -fno-tree-pta

Comment (ndkeen) [reply cime_config/machines/Depends.gnugpu.cmake:1]: Oh I see -- that was already present for gnu. So adding this flag for GNU is not new, it's just we didn't have a gnugpu Depends file yet and therefore builds with gnugpu might also need that.  I can redo PR to exclude adding those flags.

Comment (ndkeen) [reply cime_config/machines/Depends.gnugpu.cmake:1]: SO you think I should remove the flags to set `-fno-tree-pta`?

Comment (jayeshkrishna) [reply cime_config/machines/Depends.gnugpu.cmake:1]: No, just document that you are adding the flag in this PR

Comment (jayeshkrishna) [reply cime_config/machines/Depends.gnugpu.cmake:1]: ah, ok, thanks

**Review by jayeshkrishna — APPROVED** [review]

**Comment by ndkeen** [comment]
merged to next

**Pull request merged**